ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,275 @@
+from ddtrace.internal.compat import stringify
+from ddtrace.internal.utils import ArgumentError
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.internal.utils import set_argument_value
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.internal.utils.version import parse_version
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+from ddtrace.pin import Pin
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+_graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(_graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+
+config._add(
+    "graphql",
+    dict(
+        _default_service="graphql",
+        resolvers_enabled=asbool(os.getenv("DD_TRACE_GRAPHQL_RESOLVERS_ENABLED", default=False)),
+    ),
+)
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(wrap, module_str, func_name, wrapper)
+
+
+def unpatch():
+    if not getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(unwrap, module_str, func_name, wrapper)
+
+    setattr(graphql, "_datadog_patch", False)
+
+
+def _get_patching_candidates():
+    if graphql_version < (3, 0):
+        return [
+            ("graphql.graphql", "execute_graphql", _traced_query),
+            ("graphql.language.parser", "parse", _traced_parse),
+            ("graphql.validation.validation", "validate", _traced_validate),
+            ("graphql.execution.executor", "execute", _traced_execute),
+        ]
+    return [
+        ("graphql.graphql", "graphql_impl", _traced_query),
+        ("graphql.language.parser", "parse", _traced_parse),
+        ("graphql.validation.validate", "validate", _traced_validate),
+        ("graphql.execution.execute", "execute", _traced_execute),
+    ]
+
+
+def _update_patching(operation, module_str, func_name, wrapper):
+    module = sys.modules[module_str]
+    func = getattr(module, func_name)
+    operation(func, wrapper)
+
+
+def _traced_parse(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    # If graphql.parse() is called outside graphql.graphql(), graphql.parse will
+    # be a top level span. Thereforce we must explicitly set the service name.
+    with pin.tracer.trace(
+        name="graphql.parse",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ):
+        return func(*args, **kwargs)
+
+
+def _traced_validate(func, args, kwargs):
+    pin = Pin.get_from(graphql)
```

Comment:
Let's see if we can use `with_traced_module()` to remove the duplicate logic of Pin-checking

Reply:
`ddtrace.internal.wrapping.wrap()` expects a wrapper with the following signature: [wrapper(func, args, kwargs)](https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/internal/wrapping.py#L336).. `@trace_utils.with_traced_module` returns a function with the following signature:  [wrapper(func, args, INSTANCE, kwargs)](https://github.com/DataDog/dd-trace-py/blob/1.x/ddtrace/contrib/trace_utils.py#L159). Using `with_traced_module` raises a TypeError due to the missing instance parameter. 

Since `with_traced_module` is not compatible with bytecode wrapper used in the graphql integration
I can add a new decorator which supports the bytecode wrapper but I don't think we should do this in this PR.